device/core/src/device_client.ts

```diff
@@ -171,13 +175,77 @@ export class Client extends InternalClient {
+   *
+   * @throws {ReferenceException} If blobName is falsy.
+   */
+  getBlobSharedAccessSignature(blobName: string, callback: Callback<UploadParams>): void;
+  getBlobSharedAccessSignature(blobName: string): Promise<UploadParams>;
+  getBlobSharedAccessSignature(blobName: string, callback?: Callback<UploadParams>): Promise<UploadParams> | void {
+    return callbackToPromise((_callback) => {
+      /*Codes_SRS_NODE_DEVICE_CLIENT_41_XXX: [The `getBlobSharedAccessSignature` method shall throw a `ReferenceError` if `blobName` is falsy.]*/
+      if (!blobName) throw new ReferenceError('blobName cannot be \'' + blobName + '\'');
+      const retryOp = new RetryOperation(this._retryPolicy, this._maxOperationTimeout);
+      retryOp.retry((opCallback) => {
+        /*Codes_SRS_NODE_DEVICE_CLIENT_41_XXX: [The `getBlobSharedAccessSignature` method shall call the `getBlobSharedAccessSignature` method in the instantiated `_fileUploadApi` class and pass in `blobName` as a parameter.]*/
+        this._fileUploadApi.getBlobSharedAccessSignature(blobName, opCallback);
+      }, (err, result) => {
+        /*Codes_SRS_NODE_DEVICE_CLIENT_41_XXX: [The `getBlobSharedAccessSignature` method shall call the `_callback` callback with `err` and `result` from the call to `getBlobSharedAccessSignature`.]*/
+        if (!err) {
+          debug('got blob storage shared access signature.');
+          /*Codes_SRS_NODE_DEVICE_CLIENT_41_XXX: [The `getBlobSharedAccessSignature` method shall store the `result` value for use by the `notifyBlobUploadStatus` method.]*/
+          this._blobStorageUploadParams = result;
+        } else {
+          debug('Could not obtain blob shared access signature.');
+        }
+        safeCallback(_callback, err, result);
+      });
+    }, callback);
+  }
+
+  /**
+   * @description      The `notifyBlobUploadStatus` method sends IoT Hub the result of a blob upload.
+   *
+   * @param {BlobUploadCommonResponseStub}   uploadResponse   The reponse received after a blob upload via the storage api is complete
+   * @param {ErrorCallback}                  [callback]       Optional callback to call when the upload is complete.
+   * @returns {Promise<void> | void}                          Promise if no callback function was passed, void otherwise.
+   *
+   * @throws {ReferenceException} If uploadResponse is falsy.
+   */
+  notifyBlobUploadStatus(err: Error, uploadResponse: BlobUploadCommonResponseStub, callback: ErrorCallback): void;
+  notifyBlobUploadStatus(err: Error, uploadResponse: BlobUploadCommonResponseStub): Promise<void>;
+  notifyBlobUploadStatus(err: Error, uploadResponse: BlobUploadCommonResponseStub, callback?: ErrorCallback): Promise<void> | void {
+    return callbackToPromise((_callback) => {
+      /*Codes_SRS_NODE_DEVICE_CLIENT_41_XXX: [The `notifyBlobUploadStatus` method shall throw a `ReferenceError` if `uploadResponse` is falsy.]*/
```

Comment:
make sure these are numbered

Reply:
also see if these are reflected in tests

---
In reply to: [321486450](https://github.com/Azure/azure-iot-sdk-node/pull/610#discussion_r321486450) [](ancestors = 321486450)